### DISCUSSION TITLE
[Item] Literature Summary Note with metadata and links (Markdown)

### DISCUSSION THREAD

**Discussion by netgrims** [windingwind/zotero-better-notes#857] [Note Templates]
### Description

This is a template I made to take summary notes of literature. The notes are intended to be exported and used with an external markdown editor (in my case Zettlr), while including metadata from Zotero and links to the parent item and the PDF (both of which should open in Zotero).

It works for almost all item types in Zotero (ie. journal articles, conference papers, reports, books, book sections, encyclopedia articles, newspaper articles, theses, webpages, blog posts, magazine articles, documents, and dictionary entries). Information such as publisher, DOI, ISBN, etc. automatically adapt to the right item type.

Obviously, it is very much adapted to my own style of note taking. Feel free to do the same for your own needs.

## Overview of contents
- H1: citekey (I use the BetterBibTex plugin for this)
    - Reading status (to fill in yourself)
    - Note date (automatically inserted)
    - Tags that indicate that this is a literature summary note (for the external markdown editor)
- H2: title
- H2: Metadata
    - Title
    - Authors (short)
    - Publication date
    - Item type
    - Journal/publisher/publication/website/conference/...
    - Pages
    - All contributors
    - Identifier (ie. DOI, ISBN, URL)
    - Local link (Zotero item and PDF)
    - Citekey (based on BBT)
    - Citation count (based on plugin Citation Count Manager)
- H2: Abstract (automatically added)
- H2: Summary
- H2: Method
- H2: My Notes
- H2: Associations
- H2: Relevant Sources
- H2: Loose notes
    - Other notes under this parent item in Zotero get automatically inserted here.

## Instructions
- Make sure you have the Better Notes plugin and the BetterBibTex plugin installed, and that you're using version 6 of Zotero and version 1.0.4 of Better Notes.
- Open the Template Editor, click "New" and select the new blank template
- Change the name to "[Item] Literature Summary Note" (or whatever you want to call it, but it needs to start with "[Item]")
- Copy the code from below into the empty field, and click "Save". 
- Use the template as you would normally.
- If you get the error `TypeError: Zotero.Items.get(...).filter(...)[0] is undefined`, you are probably using the template on an item with no PDF attachment, which will not work. For items with no PDF attachment, use this version of the template instead: https://github.com/windingwind/zotero-better-notes/discussions/857#discussioncomment-8159391

## Issues
There are some issues that I have so far been unable to resolve with this template:
- When exporting to markdown, Better Notes will add a "\\" before many of the symbols that could be interpreted by Markdown. I work around this by using 'find' and 'replace all' for "\\" in my markdown editor.
- This template only works for items in Zotero that have a PDF attachment, due to the JS code inserting a link to the PDF. If you do try this, it will give you the following error: `TypeError: Zotero.Items.get(...).filter(...)[0] is undefined` If this is a problem for you, just remove the fragment of code for PDF linking:
```
 — [PDF](zotero://open-pdf/0_${Zotero.Items.get(topItem.getAttachments()).filter((i) => i.isPDFAttachment())[0].key})
```
- The last part of this template inserts the other child notes that are attached to the parent item. It will always insert an empty note. Also, the first heading will be lost, I think because markdown is not the default format for notes in Zotero.
- I wanted to differentiate between different kinds of contributors (ie. authors, editors, contributors, book editors, etc.), but have so far not found a way to achieve this. For example, `topItem.getField("editor")` does not work. I don't know what other key to use.

These are minor issues, which don't break the template. However, if anyone knows of a solution for any of them, feel free to comment below.

### Screenshots

![image](https://github.com/windingwind/zotero-better-notes/assets/89035122/14e304b0-7135-4292-aec3-710723a8e689)

### Test on Zotero version

6.0.30

### Test on Better Notes version

1.0.4

### Template Type

Item

### Template Share Code

```yaml
// @use-markdown
// @author [netgrims](https://github.com/netgrims)
// @link https://github.com/windingwind/zotero-better-notes/discussions/857#discussion-6061987
# @${topItem.getField('citekey')}
*Reading status*:  
*Note date: ${new Date().toISOString().substr(0,10)+" "+ new Date().toTimeString()}*  
*Tags*: #literature-summary-note #LSM  
## ${topItem.getField("title")}
---
## Metadata
- **Title**: ${topItem.getField('title')} 
- **Authors**: ${topItem.getField('firstCreator')} 
- **Publication date**: ${topItem.getField('date')} 
- **Item type**: ${topItem.itemType} 
- **${(()=>{
              if(topItem.itemType === "conferencePaper") return "Conference";
              if(topItem.itemType === "journalArticle") return "Journal";
              if(topItem.itemType === "report") return "Publisher";
              if(topItem.itemType === "book") return "Publisher";
              if(topItem.itemType === "bookSection") return "In";
              if(topItem.itemType === "encyclopediaArticle") return "In";
              if(topItem.itemType === "newspaperArticle") return "Publication";
              if(topItem.itemType === "thesis") return "Publication";
              if(topItem.itemType === "webpage") return "Website";
              if(topItem.itemType === "blogPost") return "Blog";
              if(topItem.itemType === "magazineArticle") return "Publication";
              if(topItem.itemType === "document") return "Publisher";
              if(topItem.itemType === "dictionaryEntry") return "Publisher";
              })()}**: 
${(()=>{
              if(topItem.itemType === "conferencePaper") {
                const res =  topItem.getField("conferenceName");
                return res?res:topItem.getField("proceedingsTitle");
              };
              if(topItem.itemType === "journalArticle") return topItem.getField("publicationTitle");
              if(topItem.itemType === "report") return topItem.getField("institution");
              if(topItem.itemType === "book") return topItem.getField("publisher");
              if(topItem.itemType === "bookSection") return topItem.getField("bookTitle") + " (" + topItem.getField("publisher") + ")";
              if(topItem.itemType === "encyclopediaArticle") return topItem.getField("encyclopediaTitle") + " (" +  topItem.getField("publisher") + ")";
              if(topItem.itemType === "newspaperArticle") return topItem.getField("publicationTitle");
              if(topItem.itemType === "thesis") return topItem.getField("publisher");
              if(topItem.itemType === "webpage") return topItem.getField("publicationTitle");
              if(topItem.itemType === "blogPost") return topItem.getField("publicationTitle");
              if(topItem.itemType === "magazineArticle") topItem.getField("publicationTitle");
              if(topItem.itemType === "document") return topItem.getField("publisher");
              if(topItem.itemType === "dictionaryEntry") return topItem.getField("publisher");
              })()}

- **Pages**: ${topItem.getField('pages')} 
- **All contributors**: ${topItem.getCreators().map((v)=>v.firstName+" "+v.lastName).join("; ")}
- **Identifier**: ${(() => {
        const doi = topItem.getField("DOI");
        const isbn = topItem.getField("ISBN");
        if (doi) {
          return `<a href="https://doi.org/${topItem.getField('DOI')}">${topItem.getField('DOI')}</a>`;
        } else if (isbn) {
          return `<a href="https://search.worldcat.org/search?q=${topItem.getField('ISBN')}">${topItem.getField('ISBN')}</a>`;
        } else {
          return `<a href="${topItem.getField('url')}">${topItem.getField('url')}</a>`;
        }
      })()} 
      
- **Local link**: [Zotero](zotero://select/library/items/${topItem.key}) — [PDF](zotero://open-pdf/0_${Zotero.Items.get(topItem.getAttachments()).filter((i) => i.isPDFAttachment())[0].key})

- **Citekey**: ${topItem.getField('citekey')} 

- **Citation count**: ${(() => {
        var extraField = topItem.getField('extra');
        if (extraField.split(" ").includes("citations")) {
          var citationCount = extraField.split('\n')[0];
          return citationCount;
        } else {
          return ''
        }
      })()}
---

## Abstract:
"${topItem.getField('abstractNote')}" [@${topItem.getField('citekey')}] 

---
## Summary:
### Main points

### Premise

### Literature

### Findings

### Discussion

---
## Method:
### Sampling

### Data gathering

### Data analysis

---
## My notes:
### Ideas

### Scientific significance

### Shortcomings

---
## Associations:

---
## Relevant sources:

---
### Loose notes:
${itemNotes.map((noteItem)=>{
const notePar = `
---
    ${noteItem.getNote()}
---
`;
copyNoteImage(noteItem);
return notePar;
}).join("\n")}
```

**Comment by windingwind** [comment]
Wonderful!

For the beta version of BN, you may want to replace the

```js
${itemNotes.map((noteItem)=>{
const notePar = `
---
    ${noteItem.getNote()}
---
`;
copyNoteImage(noteItem);
return notePar;
}).join("\n")}
```

with 

```js
${await Zotero.BetterNotes.api.note2html(itemNotes, { targetNoteItem })}
```

See also

https://github.com/windingwind/zotero-better-notes/blob/f976d60591428b8328ce9adf05514641ce5b1f80/src/modules/convert/api.ts#L230-L233

**Reply by netgrims** [reply]
Thanks! I'm staying on Zotero 6 for now (and I assume the beta version of BN is for Z7), so I'm going to leave it like this for now.

**Comment by Jiiron** [comment]
hi, how to fix this?
TypeError: Zotero.Items.get(...).filter(...)[0] is undefined

**Reply by netgrims** [reply]
I get the same error. However, when I use it on items in Zotero that have a PDF attachment, it is no problem, and you can use the template despite the error. On items in Zotero that don't have a PDF attachment, you can use the template below (which doesn't fetch a link to the Zotero item and the PDF). My knowledge of JavaScript is too limited to fix the issue completely.

```
// @use-markdown
// @author [netgrims](https://github.com/netgrims)
// @link https://github.com/windingwind/zotero-better-notes/discussions/857#discussion-6061987
# @${topItem.getField('citekey')}
*Reading status*:  
*Note date: ${new Date().toISOString().substr(0,10)+" "+ new Date().toTimeString()}*  
*Tags*: #literature-summary-note #LSM  
## ${topItem.getField("title")}
---
## Metadata
- **Title**: ${topItem.getField('title')} 
- **Authors**: ${topItem.getField('firstCreator')} 
- **Publication date**: ${topItem.getField('date')} 
- **Item type**: ${topItem.itemType} 
- **${(()=>{
              if(topItem.itemType === "conferencePaper") return "Conference";
              if(topItem.itemType === "journalArticle") return "Journal";
              if(topItem.itemType === "report") return "Publisher";
              if(topItem.itemType === "book") return "Publisher";
              if(topItem.itemType === "bookSection") return "In";
              if(topItem.itemType === "encyclopediaArticle") return "In";
              if(topItem.itemType === "newspaperArticle") return "Publication";
              if(topItem.itemType === "thesis") return "Publication";
              if(topItem.itemType === "webpage") return "Website";
              if(topItem.itemType === "blogPost") return "Blog";
              if(topItem.itemType === "magazineArticle") return "Publication";
              if(topItem.itemType === "document") return "Publisher";
              if(topItem.itemType === "dictionaryEntry") return "Publisher";
              })()}**: 
${(()=>{
              if(topItem.itemType === "conferencePaper") {
                const res =  topItem.getField("conferenceName");
                return res?res:topItem.getField("proceedingsTitle");
              };
              if(topItem.itemType === "journalArticle") return topItem.getField("publicationTitle");
              if(topItem.itemType === "report") return topItem.getField("institution");
              if(topItem.itemType === "book") return topItem.getField("publisher");
              if(topItem.itemType === "bookSection") return topItem.getField("bookTitle") + " (" + topItem.getField("publisher") + ")";
              if(topItem.itemType === "encyclopediaArticle") return topItem.getField("encyclopediaTitle") + " (" +  topItem.getField("publisher") + ")";
              if(topItem.itemType === "newspaperArticle") return topItem.getField("publicationTitle");
              if(topItem.itemType === "thesis") return topItem.getField("publisher");
              if(topItem.itemType === "webpage") return topItem.getField("publicationTitle");
              if(topItem.itemType === "blogPost") return topItem.getField("publicationTitle");
              if(topItem.itemType === "magazineArticle") topItem.getField("publicationTitle");
              if(topItem.itemType === "document") return topItem.getField("publisher");
              if(topItem.itemType === "dictionaryEntry") return topItem.getField("publisher");
              })()}

- **Pages**: ${topItem.getField('pages')} 
- **All contributors**: ${topItem.getCreators().map((v)=>v.firstName+" "+v.lastName).join("; ")}
- **Identifier**: ${(() => {
        const doi = topItem.getField("DOI");
        const isbn = topItem.getField("ISBN");
        if (doi) {
          return `<a href="https://doi.org/${topItem.getField('DOI')}">${topItem.getField('DOI')}</a>`;
        } else if (isbn) {
          return `<a href="https://search.worldcat.org/search?q=${topItem.getField('ISBN')}">${topItem.getField('ISBN')}</a>`;
        } else {
          return `<a href="${topItem.getField('url')}">${topItem.getField('url')}</a>`;
        }
      })()} 
      
- **Local link**: [Zotero](zotero://select/library/items/${topItem.key})

- **Citekey**: ${topItem.getField('citekey')} 

- **Citation count**: ${(() => {
        var extraField = topItem.getField('extra');
        if (extraField.split(" ").includes("citations")) {
          var citationCount = extraField.split('\n')[0];
          return citationCount;
        } else {
          return ''
        }
      })()}
---

## Abstract:
"${topItem.getField('abstractNote')}" [@${topItem.getField('citekey')}] 

---
## Summary:
### Main points

### Premise

### Literature

### Findings

### Discussion

---
## Method:
### Sampling

### Data gathering

### Data analysis

---
## My notes:
### Ideas

### Scientific significance

### Shortcomings

---
## Associations:

---
## Relevant sources:

---
### Loose notes:
${itemNotes.map((noteItem)=>{
const notePar = `
---
    ${noteItem.getNote()}
---
`;
copyNoteImage(noteItem);
return notePar;
}).join("\n")}
```

**Reply by Jiiron** [reply]
this one is quite nice too, simple and clear.  But the previous one still doesn't work in my Zotero, no matte with or without PDF files

**Reply by netgrims** [reply]
Thanks!

Odd though that it won't work for you. It works for me.

In any case, I might look into replacing the code for the PDF link at some point, which would hopefully solve the problem. Although only having a link to the Zotero item is actually functionally nearly the same.

**Comment by netgrims** [comment]
Small update: I shortened the code to fetch a link to the local zotero item. It should now work for items without a PDF attachment as well, and now looks like this:
```
zotero://select/items/0_${topItem.key}
```
I tried implementing the "newer" format of `zotero://select/library/items/${topItem.key}`, but that didn't work for me. Zotero just freezes.

**Reply by netgrims** [reply]
Actually, never mind that. I forgot to delete the "0_" that was in the old link format. It works with the new format, and I updated the template above.

It now looks like this:
`zotero://select/library/items/${topItem.key}`

**Reply by netgrims** [reply]
Also, just to be clear, using the full template on Zotero items with no PDF attachment is still not working, as it will produce this error: `TypeError: Zotero.Items.get(...).filter(...)[0] is undefined`

To use it on Zotero items with no attachment, either remove the fragment that fetches the PDF link (see first post above) or copy the code from my comment above (https://github.com/windingwind/zotero-better-notes/discussions/857#discussioncomment-8159391)

**Reply by netgrims** [reply]
I suppose this could help to resolve this issue further: https://github.com/windingwind/zotero-better-notes/issues/534#issuecomment-1609274439

**Comment by TreeWoman** [comment]
New to templates in Zotero...every time I have attempted to use this, I get an: 

**Template [Literature Review] Error: ReferenceError: topItem is not defined**

It _looks_ as though this refers to the cite - and BibTex is installed and generating a key - but, it still does absolutely nothing. 

Realize this must be my error - but don't really understand how to fix it. Any help you can provide would be very appreciated.

**Reply by netgrims** [reply]
Did you name the template with "[Item]" in the beginning of the name? It sounds like that might be the issue. And just to be sure: you have the better-notes plugin installed, right?

**Reply by TreeWoman** [reply]
Thanks for your quick response - it is very appreciated.

Yes, Better Notes is installed and working fine.

I didn't see any instruction to rename the template..although I did rename it to [LiteratureTemplate] in the Template Editor. So - was it supposed to be given a different name..or not? I am using the template "share code" and copy/pasting it into the Template Editor in Zotero and then saving it. 

I just repeated the steps again - and it is still throwing the same error.

**Reply by netgrims** [reply]
Oh, then that's probably the issue. The name needs to start with "[Item]". So, for example, you could call it "[Item] Literature Template". This is just so that the software knows that it is an item template. You can just rename the template in the template editor and then it should work.

**Reply by TreeWoman** [reply]
That did it! Completely awesome...! THANK YOU so much!!!!!

**Reply by netgrims** [reply]
No problem! It's a bit confusing, I know. Anyway, I'm happy to hear you're liking it.

**Comment by TreeWoman** [comment]
I am having some difficulty with BetterNotes being used by the Literature Template I installed into Zotero. While simple info is coming through fine...when it comes to BetterNotes - I am getting this:

![image](https://github.com/windingwind/zotero-better-notes/assets/55290005/167368d8-43ea-49a4-a390-17f67936018d)

Have I missed a setting? Or is the Lit Review template not supposed to import the relevant BetterNotes? Thanks in advance!
